src/Middleware/HttpLogging/src/FileLoggerProcessor.cs

```diff
@@ -0,0 +1,163 @@
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal class FileLoggerProcessor : IDisposable
+    {
+        private const int _maxQueuedMessages = 1024;
+
+        private readonly string _path;
+        private readonly string _fileName;
+        private readonly int? _maxFileSize;
+        private readonly int? _maxRetainedFiles;
+        private int _fileNumber = 1;
+
+        private readonly BlockingCollection<LogMessage> _messageQueue = new BlockingCollection<LogMessage>(_maxQueuedMessages);
+        private readonly List<LogMessage> _currentBatch = new List<LogMessage>();
+        private Task _outputTask;
+        private CancellationTokenSource _cancellationTokenSource;
+
+        public FileLoggerProcessor(IOptionsMonitor<FileLoggerOptions> options)
+        {
+            var loggerOptions = options.CurrentValue;
+            _path = loggerOptions.LogDirectory;
+            _fileName = loggerOptions.FileName;
+            _maxFileSize = loggerOptions.FileSizeLimit;
+            _maxRetainedFiles = loggerOptions.RetainedFileCountLimit;
+
+            // Start message queue processor
+            _cancellationTokenSource = new CancellationTokenSource();
+            _outputTask = Task.Run(ProcessLogQueue);
+        }
+
+        public void EnqueueMessage(LogMessage message)
+        {
+            if (!_messageQueue.IsAddingCompleted)
+            {
+                try
+                {
+                    _messageQueue.Add(message);
+                    return;
+                }
+                catch (InvalidOperationException) { }
+            }
+        }
+
+        private async Task ProcessLogQueue()
+        {
+            while (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                while (_messageQueue.TryTake(out var message))
+                {
+                    _currentBatch.Add(message);
+                }
+                if (_currentBatch.Count > 0)
+                {
+                    try
+                    {
+                        await WriteMessagesAsync(_currentBatch);
+                    }
+                    catch
+                    {
+                        // ignored
+                    }
+
+                    _currentBatch.Clear();
+                }
+            }
+        }
+
+        internal async Task WriteMessagesAsync(IEnumerable<LogMessage> messages)
+        {
+            Directory.CreateDirectory(_path);
+
+            // Files are grouped by day, and written up to _maxFileSize before rolling to a new file
+            foreach (var group in messages.GroupBy(GetGrouping))
+            {
+                var fullName = GetFullName(group.Key);
+                var fileInfo = new FileInfo(fullName);
+                var streamWriter = File.AppendText(fullName);
```

Comment:
using var

Reply:
Should also probably open this file once and keep writing to it until you roll to a new file

Reply:
I didn't use `using` because if the file reaches max_length I need to close the streamWriter and re-open it on a new file - I believe this is already doing what you describe (per day, I open a streamWriter on file 1. Once file 1 is full, I dispose the streamWriter and re-open it on file 2, etc).

Reply:
You're opening the file per "group" of messages right now. And the groups are just whatever logs are in the collection when you iterate over it. I'm saying you should get rid of the grouping, open the file the first time it's needed and keep it open (with the checks for it being full, new day, etc.).

Reply:
I got rid of the grouping, but I think I still can't do `using`, since inside that using block I might need to roll to a new file if the current one reaches `max_file_size`